format/postgres/common/pg_btree/postgres/pg_btree.go

```diff
@@ -0,0 +1,284 @@
+package postgres
+
+import (
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/format/postgres/common"
+	"github.com/wader/fq/format/postgres/common/pg_heap/postgres"
+	"github.com/wader/fq/pkg/decode"
+)
+
+const (
+	BTREE_MAGIC = 0x053162
+	P_NONE      = 0
+
+	/* Bits defined in btpo_flags */
+	BTP_LEAF             = 1 << 0 /* leaf page, i.e. not internal page */
+	BTP_ROOT             = 1 << 1 /* root page (has no parent) */
+	BTP_DELETED          = 1 << 2 /* page has been deleted from tree */
+	BTP_META             = 1 << 3 /* meta-page */
+	BTP_HALF_DEAD        = 1 << 4 /* empty, but still in tree */
+	BTP_SPLIT_END        = 1 << 5 /* rightmost page of split group */
+	BTP_HAS_GARBAGE      = 1 << 6 /* page has LP_DEAD tuples (deprecated) */
+	BTP_INCOMPLETE_SPLIT = 1 << 7 /* right sibling's downlink is missing */
+	BTP_HAS_FULLXID      = 1 << 8 /* contains BTDeletedPageData */
+)
+
+const (
+	INDEX_SIZE_MASK       = 0x1FFF
+	INDEX_AM_RESERVED_BIT = 0x2000 /* reserved for index-AM specific usage */
+	INDEX_VAR_MASK        = 0x4000
+	INDEX_NULL_MASK       = 0x8000
+)
+
+const (
+	IndexTupleDataSize = 8
+)
+
+// struct BTMetaPageData {
+/*    0      |     4 */ // uint32 btm_magic
+/*    4      |     4 */ // uint32 btm_version
+/*    8      |     4 */ // BlockNumber btm_root
+/*   12      |     4 */ // uint32 btm_level
+/*   16      |     4 */ // BlockNumber btm_fastroot
+/*   20      |     4 */ // uint32 btm_fastlevel
+/*   24      |     4 */ // uint32 btm_last_cleanup_num_delpages
+/* XXX  4-byte hole */
+/*   32      |     8 */ // float8 btm_last_cleanup_num_heap_tuples
+/*   40      |     1 */ // _Bool btm_allequalimage
+/* XXX  7-byte padding */
+//
+/* total size (bytes):   48 */
+
+// struct BTPageOpaqueData {
+/*    0      |     4 */ // BlockNumber btpo_prev;
+/*    4      |     4 */ // BlockNumber btpo_next;
+/*    8      |     4 */ // uint32 btpo_level;
+/*   12      |     2 */ // uint16 btpo_flags;
+/*   14      |     2 */ // BTCycleId btpo_cycleid;
+//
+/* total size (bytes):   16 */
+
+// struct IndexTupleData {
+/*    0      |     6 */ // ItemPointerData t_tid;
+/*    6      |     2 */ // unsigned short t_info;
+//
+// IndexTupleData *IndexTuple;
+/* total size (bytes):    8 */
+
+func DecodePgBTree(d *decode.D, args format.PostgresBTreeIn) any {
+	btree := &BTree{
+		Args:     args,
+		PageSize: common.PageSize,
+	}
+	decodeBTreePages(btree, d)
+	return nil
+}
+
+type BTree struct {
+	Args     format.PostgresBTreeIn
+	PageSize uint64
+	page     *postgres.HeapPage
+}
+
+func decodeBTreePages(btree *BTree, d *decode.D) {
+	for i := btree.Args.Page; ; i++ {
+		page := &postgres.HeapPage{}
+		if btree.page != nil {
+			// use prev page
+			page.BytesPosBegin = btree.page.BytesPosEnd
+		}
+		page.BytesPosEnd = int64(common.TypeAlign(btree.PageSize, uint64(page.BytesPosBegin)+1))
+		btree.page = page
+
+		pos0 := page.BytesPosBegin * 8
+		d.SeekAbs(pos0)
+
+		if end, _ := d.TryEnd(); end {
```

Comment:
I think `d.SeekAbs` will panic and abort decode if pos is invalid so maybe can just be `if d.End()`? or was there some reason `TryEnd` was used?

Reply:
It's done.